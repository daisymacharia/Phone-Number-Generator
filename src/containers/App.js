import React, { Component } from 'react'
import { inject, observer } from 'mobx-react'
import styled from 'styled-components'

const Page = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
  background: #f5f5f5;
  position: absolute;
  width: 100%;
  height: 100%;
  padding: 5rem;
  box-sizing: border-box;
`

const StyledInputContainer = styled.div`
  box-shadow: inset 0 -2px 1px rgba(0, 0, 0, 0.09);
  background: #ffff;
  width: 35rem;
  padding: 20px 10px 10px 20px;
`

const StyledInput = styled.input`
  margin: 0;
  width: 100%;
  font-size: 24px;
  border: none;
  outline: none;
  font-family: Helvetica Neue, Helvetica, Arial, sans-serif;
  font-size: 30px;
  font-weight: 100;
`

const StyledButton = styled.button`
  width: 10rem;
  height: 3rem;
  margin: 3px;
  padding: 3px 7px;
  text-decoration: none;
  border: 1px solid #ece9e9;
  border-radius: 3px;
  outline: none;
  cursor: pointer;
`

const Container = styled.div`
  display: flex;
  padding: 10px;
  min-height: fit-content;
`
const Span = styled.span`
  font-size: ${props => props.size || '16px'};
  font-family: Helvetica Neue, Helvetica, Arial, sans-serif;
  font-weight: ${props => props.weight || 200};
  margin-bottom: 10px;
`
const Alerts = styled.div`
  display: flex;
  flex-direction: column-reverse;
  position: absolute;
  top: 0;
  right: 0;
`

const SingleAlert = styled.div`
  display: flex;
  min-width: 400px;
  position: relative;
  background: #f5f5f5;
  box-shadow: 0 0 5px black;
  margin-bottom: 15px;
  padding: 10px;
  border-radius: 4px;
  animation: fadeOut 4s linear forwards;

  @keyframes fadeOut {
    0% {
      opacity: 0;
    }
    10% {
      opacity: 1;
    }
    90% {
      opacity: 1;
      transform: translateY(0px);
    }
    100% {
      opacity: 0;
      transform: translateY(50px);
    }
  }
`

@inject('store')
@observer
class App extends Component {
  state = {
    limit: 0,
  }

  generateNumbers = () => {
    return (
      (this.props.store.limit = this.state.limit),
      this.props.store.generateRandomNumbers()
    )
  }

  isNumber = evt => {
    var charCode = evt.which ? evt.which : evt.keyCode
    if (charCode > 31 && (charCode < 48 || charCode > 57))
      return this.props.store.notNumber()

    return this.setState({ limit: evt.target.value })
  }

  render() {
    const { store } = this.props

    return (
      <Page>
        <Span size='40px' weight={500}>
          Random Number Generator
        </Span>
        <Container>
          <StyledInputContainer>
            <StyledInput
              onKeyUp={this.isNumber}
              placeholder='Number of phone numbers to be generated'
            />
          </StyledInputContainer>
          <StyledButton onClick={this.generateNumbers}>
            Generate Numbers
          </StyledButton>
        </Container>

        {store.phoneNumbers.length > 0 && (
          <>
            <Container>
              <StyledButton
                onClick={() =>
                  store.saveNumbers(store.phoneNumbers, 'unsorted.csv')
                }>
                Download unsorted Numbers
              </StyledButton>

              <StyledButton
                onClick={() =>
                  store.saveNumbers(store.sortedNumbers, 'sorted.csv')
                }>
                Download sorted Numbers
              </StyledButton>

              <StyledButton onClick={() => store.sortNumbers('asc')}>
                Sort in Ascending
              </StyledButton>
              <StyledButton onClick={() => store.sortNumbers('desc')}>
                Sort in Descending
              </StyledButton>
            </Container>
            {store.sortedNumbers.length > 0 && (
              <Span>Phone Numbers sorted</Span>
            )}
            <Span>Total Number of Generated Numbers: {store.limit}</Span>
            <Span>The Max Number is: {store.MaxNumber}</Span>
            <Span>The Min Number is: {store.MinNumber}</Span>
          </>
        )}
        {store.messages.length > 0 && (
          <Alerts>
            {store.messages.map((message, idx) => (
              <SingleAlert key={idx + message}>
                <p> {message} </p>
              </SingleAlert>
            ))}
          </Alerts>
        )}
      </Page>
    )
  }
}
export default App
