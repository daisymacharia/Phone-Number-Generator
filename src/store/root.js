import { observable, action, computed } from 'mobx'
import { saveAs } from 'file-saver'

export default class Store {
  @observable phoneNumbers = []
  @observable sortedNumbers = []
  @observable limit = 0
  @observable MinNumber = 0
  @observable MaxNumber = 0
  @observable message = ''

  @action
  generateRandomNumbers = () => {
    var numbers = new Array(this.limit)
    if (this.limit >= 10000) {
      this.message = 'You can only generate upto 10000 numbers at a time'
    } else {
      for (var i = 0; i < this.limit; i++) {
        numbers[i] = this.addZero(Math.floor(Math.random() * 1000000000))
      }

      this.phoneNumbers = numbers
      this.getMaxMinNumbers()
      return this.phoneNumbers
    }
  }

  addZero = number => {
    var str = `${number}`
    var length = 10
    while (str.length < length) {
      str = `${0 + str}`
    }
    return str
  }

  @action
  saveNumbers(numbers, fileName) {
    saveAs(new Blob([numbers], { type: 'text/csv;charset=utf-8' }), fileName)
  }

  getMaxMinNumbers() {
    this.MinNumber = this.phoneNumbers.slice().sort((a, b) => a - b)[0]
    this.MaxNumber = this.phoneNumbers
      .slice()
      .sort((a, b) => a - b)
      .reverse()[0]
  }

  @action
  sortNumbers = order => {
    this.sortedNumbers = this.phoneNumbers.slice().sort((a, b) => a - b)

    if (order === 'asc') {
      return (
        this.sortedNumbers, (this.message = 'Numbers sorted in ascending order')
      )
    } else if (order === 'desc') {
      return (
        (this.sortedNumbers = this.sortedNumbers.slice().reverse()),
        (this.message = 'Numbers sorted in descending order')
      )
    }
  }
}
