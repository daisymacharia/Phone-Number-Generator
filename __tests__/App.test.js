import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { observable } from 'mobx'
import App from '../src/containers/App'

import * as Store from '../src/store/root'

jest.mock('../src/store/root', () => ({
  generateRandomNumbers: jest.fn(),
  sortNumbers: jest.fn(),
  saveNumbers: jest.fn(),
  notNumber: jest.fn(),
}))

describe('Random Number Generator', () => {
  let store

  //don't use an arrow function...preserve the value of "this"
  beforeEach(() => {
    store = observable({
      phoneNumbers: [
        '0516748992',
        '0806745880',
        '0428180874',
        '0571920472',
        '0454586248',
      ],
      limit: 5,
      sortedNumbers: [
        '0806745880',
        '0571920472',
        '0516748992',
        '0454586248',
        '0428180874',
      ],
      generateRandomNumbers: Store.generateRandomNumbers,
      messages: [],
      sortNumbers: Store.sortNumbers,
      saveNumbers: Store.saveNumbers,
      notNumber: Store.notNumber,
    })
  })

  it('renders correctly', function() {
    const tree = renderer.create(<App store={store} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('calls generatePhoneNumbers on generate numbers button click', function() {
    const wrapper = shallow(<App store={store} />).dive()
    wrapper
      .find('App__StyledButton')
      .at(0)
      .simulate('click')

    expect(store.generateRandomNumbers).toHaveBeenCalled()
  })

  it('saves unsorted numbers on button click', function() {
    const wrapper = shallow(<App store={store} />).dive()
    wrapper
      .find('App__StyledButton')
      .at(1)
      .simulate('click')

    expect(store.saveNumbers).toHaveBeenCalledWith(
      store.phoneNumbers,
      'unsorted.csv'
    )
  })

  it('saves sorted numbers on button click', function() {
    const wrapper = shallow(<App store={store} />).dive()
    wrapper
      .find('App__StyledButton')
      .at(2)
      .simulate('click')

    expect(store.saveNumbers).toHaveBeenCalledWith(
      store.sortedNumbers,
      'sorted.csv'
    )
  })

  it('sorts in asc on sortNumbers button click', function() {
    const wrapper = shallow(<App store={store} />).dive()
    wrapper
      .find('App__StyledButton')
      .at(3)
      .simulate('click')

    expect(store.sortNumbers).toHaveBeenCalledWith('asc')
  })

  it('sorts in desc on sortNumbers button click', function() {
    const wrapper = shallow(<App store={store} />).dive()
    wrapper
      .find('App__StyledButton')
      .at(4)
      .simulate('click')

    expect(store.sortNumbers).toHaveBeenCalledWith('desc')
  })

  it('displays generated number statistics', function() {
    const wrapper = shallow(<App store={store} />).dive()
    expect(
      wrapper
        .find('App__Span')
        .at(2)
        .text()
    ).toEqual('Total Number of Generated Numbers: 5')
    expect(
      wrapper
        .find('App__Span')
        .at(3)
        .text()
    ).toEqual('The Max Number is: ')
    expect(
      wrapper
        .find('App__Span')
        .at(4)
        .text()
    ).toEqual('The Min Number is: ')
  })

  it('alerts when inputing non integers', function() {
    const wrapper = shallow(<App store={store} />).dive()
    const input = wrapper.find('App__StyledInput')
    input.simulate('keyUp', { keyCode: 40 })
    expect(store.notNumber).toHaveBeenCalled()
  })
})
