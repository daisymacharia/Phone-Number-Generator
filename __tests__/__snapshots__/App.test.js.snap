// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Random Number Generator renders correctly 1`] = `
<div
  className="App__Page-sc-18h9djc-0 jwDQHH"
>
  <span
    className="App__Span-sc-18h9djc-5 cspJHA"
    size="40px"
  >
    Random Number Generator
  </span>
  <div
    className="App__Container-sc-18h9djc-4 gRDfrI"
  >
    <div
      className="App__StyledInputContainer-sc-18h9djc-1 ezwnbl"
    >
      <input
        className="App__StyledInput-sc-18h9djc-2 hZCLZD"
        onKeyUp={[Function]}
        placeholder="Number of phone numbers to be generated"
      />
    </div>
    <button
      className="App__StyledButton-sc-18h9djc-3 hqCJki"
      onClick={[Function]}
    >
      Generate Numbers
    </button>
  </div>
  <div
    className="App__Container-sc-18h9djc-4 gRDfrI"
  >
    <button
      className="App__StyledButton-sc-18h9djc-3 hqCJki"
      onClick={[Function]}
    >
      Download unsorted Numbers
    </button>
    <button
      className="App__StyledButton-sc-18h9djc-3 hqCJki"
      onClick={[Function]}
    >
      Download sorted Numbers
    </button>
    <button
      className="App__StyledButton-sc-18h9djc-3 hqCJki"
      onClick={[Function]}
    >
      Sort in Ascending
    </button>
    <button
      className="App__StyledButton-sc-18h9djc-3 hqCJki"
      onClick={[Function]}
    >
      Sort in Descending
    </button>
  </div>
  <span
    className="App__Span-sc-18h9djc-5 heoFad"
  >
    Phone Numbers sorted
  </span>
  <span
    className="App__Span-sc-18h9djc-5 heoFad"
  >
    Total Number of Generated Numbers: 
    5
  </span>
  <span
    className="App__Span-sc-18h9djc-5 heoFad"
  >
    The Max Number is: 
  </span>
  <span
    className="App__Span-sc-18h9djc-5 heoFad"
  >
    The Min Number is: 
  </span>
</div>
`;
